typed-protocols/src/Network/TypedProtocol/Codec.hs

```diff
@@ -0,0 +1,148 @@
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE PolyKinds                  #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE NamedFieldPuns             #-}
+{-# LANGUAGE TypeInType                 #-}
+
+module Network.TypedProtocol.Codec (
+    Codec(..)
+  , PeerKind(..)
+  , PeerHasAgency(..)
+  , WeHaveAgency
+  , TheyHaveAgency
+  , SomeMessage(..)
+  , DecodeStep(..)
+  , transformCodec
+  , cborCodec
+  ) where
+
+import           Network.TypedProtocol.Core
+                   ( Protocol(..), PeerKind(..)
+                   , PeerHasAgency(..), WeHaveAgency, TheyHaveAgency )
+
+import           Control.Monad.ST (ST)
+import           Control.Monad.Class.MonadST
+
+import qualified Codec.CBOR.Encoding as CBOR (Encoding)
+import qualified Codec.CBOR.Read     as CBOR
+import qualified Codec.CBOR.Decoding as CBOR (Decoder)
+import qualified Codec.CBOR.Write    as CBOR
+
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Builder as BS
+import qualified Data.ByteString.Builder.Extra as BS
+import qualified Data.ByteString.Lazy.Internal as LBS (smallChunkSize)
+import qualified Data.ByteString.Lazy          as LBS
+
+data Codec ps (pk :: PeerKind) failure m bytes = Codec {
+       encode :: forall (st :: ps) (st' :: ps).
+                 WeHaveAgency pk st
+              -> Message ps st st'
+              -> bytes,
+
+       decode :: forall (st :: ps).
+                 TheyHaveAgency pk st
+              -> m (DecodeStep bytes failure m (SomeMessage st))
+     }
+
+transformCodec
+  :: Functor m
+  => (bytes  -> bytes')
+  -> (bytes' -> bytes)
+  -> Codec ps pk failure m bytes
+  -> Codec ps pk failure m bytes'
+transformCodec to from Codec {encode, decode} = Codec {
+    encode = \stok msg -> to (encode stok msg),
+    decode = fmap (transformDecodeStep to from) . decode
+  }
+
+-- The types here are pretty fancy. The decode is polymorphic in the protocol
+-- state, but only for kinds that are the same kind as the protocol state.
+-- The StateToken is a type family that resolves to a singleton, and the
```

Comment:
We don't have `StateToken` data family any more, so this needs to be updated.

Reply:
Yes, I should extend and updated the codec docs as a whole.